src/building/bootstrapping.md

```diff
@@ -122,50 +124,99 @@ contribution [here][bootstrap-build].
 
 ## Understanding stages of bootstrap
 
-This is a detailed look into the separate bootstrap stages. When running
-`x.py` you will see output such as:
-
-```txt
-Building stage0 std artifacts
-Copying stage0 std from stage0
-Building stage0 compiler artifacts
-Copying stage0 rustc from stage0
-Building LLVM for x86_64-apple-darwin
-Building stage0 codegen artifacts
-Assembling stage1 compiler
-Building stage1 std artifacts
-Copying stage1 std from stage1
-Building stage1 compiler artifacts
-Copying stage1 rustc from stage1
-Building stage1 codegen artifacts
-Assembling stage2 compiler
-Uplifting stage1 std
-Copying stage2 std from stage1
-Generating unstable book md files
-Building stage0 tool unstable-book-gen
-Building stage0 tool rustbook
-Documenting standalone
-Building rustdoc for stage2
-Documenting book redirect pages
-Documenting stage2 std
-Building rustdoc for stage1
-Documenting stage2 whitelisted compiler
-Documenting stage2 compiler
-Documenting stage2 rustdoc
-Documenting error index
-Uplifting stage1 rustc
-Copying stage2 rustc from stage1
-Building stage2 tool error_index_generator
-```
+### Overview
 
-A deeper look into `x.py`'s phases can be seen here:
+This is a detailed look into the separate bootstrap stages.
+
+The convention `x.py` uses is that:
+- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
+- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
+  process is called _uplifting_.
+
+For example, `x.py build --stage 0` means to build with the beta `rustc`, and
+`test --stage 0` isn't really meaningful (it's not running tests on your changes,
+but on `beta`, so it always fails). Similarly, `doc --stage 0` means to
```

Comment:
Ok I definitely said this confusingly - `test --stage 0 library/std` is meaningful, but `test --stage 0 src/test/ui` is not (and `test` includes `src/test/ui`). I'll change it.

Reply:
I added both `test library/std` and `test src/test/ui` as examples.

Reply:
I do feel this kind of shows that the 'run stage N' model is incomplete though - why does `test --stage 0 src/test/ui`  build the compiler but not `test --stage 0 library/std`?

Reply:
I think this is just stage 0 being wonky with building things it really doesn't need, which is a bug.

Reply:
I said that wrong, sorry - I meant why does src/test/ui test the stage0 _compiler_ when library/std tests the stage0 _artifacts_.

Reply:
I think this is the source of a lot of confusion actually - `build --stage 0 library/std` and `test --stage 0 library/std` refer to the same code, but `build --stage 0 compiler/rustc` and `test --stage 0 src/test/ui` do not. But then it gets even more confusing - `test --stage 0 compiler/rustc` _will_ build the compiler 😕

Reply:
> test --stage 0 compiler/rustc will build the compiler

Sure, and it should, because this is running the compiler unit tests. That's just like how `--stage0 library/std` needs to build the library to run unit tests.

But `test --stage 0 src/test/ui`  should run the UI test suite on the bootstrap compiler... which is basically pointless? It should not build the compiler, sure, but it probably should just error out for not being supported.

Reply:
> uild --stage 0 library/std and test --stage 0 library/std refer to the same code, but build --stage 0 compiler/rustc and test --stage 0 src/test/ui do not. But then it gets even more confusing - test --stage 0 compiler/rustc will build the compiler confused

To hopefully deconfuse a few things, note that `build --stage 0 compiler/rustc` and `test --stage 0 compiler/rustc` *do* refer to the same code, just like for `library/std`. Only `test src/test/ui` is different, because it *tests the compiler it is run with*, whereas for `build` and crate-level `test` targets, they *test the thing they build*.

Reply:
Ok, I think I understand: `test compiler/rustc` is running `cargo test` on the rustc package, which has no unit/integration tests. But `test src/test/ui` is running `compiletest`, which uses the compiler currently running.

I still think this is confusing, but I'll add a bit more explaining it.

Reply:
Right. `test compiler/rustc` tests the compiler *as a regular crate*, while `test src/test/ui` does way more -- it actually needs a libstd built with the *to-be-tested* compiler and then *runs it*.